vanilla-tests/src/test/java/fixtures/discriminatorflattening/DiscriminatorTests.java

```diff
@@ -22,18 +27,40 @@ public void serializationOnDiscriminator() throws IOException {
         SerializerAdapter adapter = JacksonAdapter.createDefaultSerializerAdapter();
 
         MetricAlertResource metricAlert = new MetricAlertResource();
-        metricAlert.setCriteria(new MetricAlertSingleResourceMultipleMetricCriteria());
+        metricAlert.setCriteria(new MetricAlertSingleResourceMultipleMetricCriteria().setAdditionalProperties(Collections.singletonMap("key.1", "value.1")));
         String metricAlertJson = adapter.serialize(metricAlert, SerializerEncoding.JSON);
         verifyODataTypeInJson(metricAlertJson);
 
         MetricAlertResource metricAlert2 = adapter.deserialize(metricAlertJson, MetricAlertResource.class, SerializerEncoding.JSON);
         Assert.assertTrue(metricAlert2.getCriteria() instanceof MetricAlertSingleResourceMultipleMetricCriteria);
+        Assert.assertTrue(metricAlert2.getCriteria().getAdditionalProperties().containsKey("key.1"));
+    }
+
+    @Test
+    public void serializationOnNestedFlatten() throws IOException {
+        SerializerAdapter adapter = JacksonAdapter.createDefaultSerializerAdapter();
+
+        VirtualMachineScaleSet vmss = new VirtualMachineScaleSet()
+                .setVirtualMachineProfile(new VirtualMachineScaleSetVMProfile()
+                        .setNetworkProfile(new VirtualMachineScaleSetNetworkProfile()
+                                .setNetworkInterfaceConfigurations(Collections.singletonList(new VirtualMachineScaleSetNetworkConfiguration()
+                                        .setName("name").setPrimary(true)))));
```

Comment:
@alzimmermsft 

Appears `@JsonFlatten` on field cannot handle nested properties.
Here `VirtualMachineScaleSetNetworkConfiguration` also have some fields been flattened, but it seems not serialized correctly.

`{"properties":{"virtualMachineProfile":{"networkProfile":{"networkInterfaceConfigurations":[{"name":"name","properties.primary":true}]}}}}`
where `properties.primary` part should be `primary` under `properties`.

Reply:
Thanks for identifying this @weidongxu-microsoft, I'll investigate and submit a fix.